source/boundary_composition/function.cc

```diff
@@ -0,0 +1,205 @@
+    boundary_composition (const types::boundary_id /*boundary_indicator*/,
+                          const Point<dim> &position, const unsigned int n_comp) const
+    {
+      if (coordinate_system == ::aspect::Utilities::Coordinates::cartesian)
+        {
+          return function->value(position,n_comp);
+        }
+      else if (coordinate_system == ::aspect::Utilities::Coordinates::spherical)
+        {
+          const std_cxx11::array<double,dim> spherical_coordinates =
+            aspect::Utilities::Coordinates::cartesian_to_spherical_coordinates(position);
+          Point<dim> point;
+
+          for (unsigned int i = 0; i<dim; ++i)
+            point[i] = spherical_coordinates[i];
+
+          return function->value(point,n_comp);
+        }
+      else if (coordinate_system == Utilities::Coordinates::depth)
+        {
+          const double depth = this->get_geometry_model().depth(position);
+          Point<dim> point;
+          point(0) = depth;
+          return function->value(point,n_comp);
+        }
+      else
+        {
+          AssertThrow(false, ExcNotImplemented());
+          return numbers::signaling_nan<double>();
+        }
+    }
+
+
+    template <int dim>
+    void
+    Function<dim>::update()
+    {
+      // we get time passed as seconds (always) but may want
+      // to reinterpret it in years
+      if (this->convert_output_to_years())
+        function->set_time (this->get_time() / year_in_seconds);
+      else
+        function->set_time (this->get_time());
+    }
+
+
+
+    template <int dim>
+    double
+    Function<dim>::
+    minimal_composition (const std::set<types::boundary_id> &) const
+    {
+      return min_composition;
+    }
+
+
+
+    template <int dim>
+    double
+    Function<dim>::
+    maximal_composition (const std::set<types::boundary_id> &) const
+    {
+      return max_composition;
+    }
+
+
+    template <int dim>
+    void
+    Function<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Boundary composition model");
+      {
+        prm.enter_subsection("Function");
+        {
+          prm.declare_entry ("Coordinate system", "cartesian",
+                             Patterns::Selection ("cartesian|spherical|depth"),
+                             "A selection that determines the assumed coordinate "
+                             "system for the function variables. Allowed values "
+                             "are 'cartesian', 'spherical', and 'depth'. 'spherical' coordinates "
+                             "are interpreted as r,phi or r,phi,theta in 2D/3D "
+                             "respectively with theta being the polar angle. 'depth' "
+                             "will create a function, in which only the first "
+                             "parameter is non-zero, which is interpreted to "
+                             "be the depth of the point.");
+
+          Functions::ParsedFunction<dim>::declare_parameters (prm, 1);
```

Comment:
this is wrong -- it will only work for a single compositional field. but you want this to work also for multiple fields.